Specifications/QIR/Data-Types.md

```diff
@@ -170,49 +168,36 @@ big integers.
-copy of an existing tuple to be created.
-In QIR, this is represented by creating a new copy of the existing tuple and then
-modifying the newly-created tuple in place.
-If the compiler knows that the existing tuple is not used after the modification,
-it is possible to avoid the copy and modify the existing tuple in place.
+Many languages provide immutable tuples, along with operators that allow a modified copy of an existing tuple to be created.
+QIR supports this by requiring the runtime to track and be able to access the following given a `%Tuple*`:
+- The size of the tuple in bytes
+- The user count indicating how many handles to the tuple exist in the source code
 
-The following utility functions are provided by the classical runtime to support
-tuples and user-defined types:
+The language specific compiler is responsible for injecting calls to increase and decrease the user count as needed, as well as to accurately reflect when references to the LLVM structure representing a tuple are created and removed. 
+See the section [above](#reference-and-access-counting) regarding the distinction between access and reference counting. 
+
+In the case where the source language treats tuples as immutable values, the language-specific compiler is expected to request the necessary copies prior to modifying the tuple in place. 
+This is done by invoking the runtime function `__quantum__rt__tuple_copy` to create a byte-by-byte copy of a tuple. Unless the copying is forced via the second argument, the runtime may omit copying the value and instead simply return a pointer to the given argument if the user count is 0 and it is hence save to modify the tuple in place.
+
+The following utility functions are provided by the classical runtime to support tuples and user-defined types:
 
 | Function                         | Signature             | Description |
 |----------------------------------|-----------------------|-------------|
-| __quantum__rt__tuple_create      | `%TupleHeader*(i64)`  | Allocates space for a tuple requiring the given number of bytes and sets the reference count to 1. |
-| __quantum__rt__tuple_reference   | `void(%TupleHeader*)` | Indicates that a new reference has been added. |
-| __quantum__rt__tuple_unreference | `void(%TupleHeader*)` | Indicates that an existing reference has been removed and potentially releases the tuple. |
+| __quantum__rt__tuple_create      | `%Tuple*(i64)`  | Allocates space for a tuple requiring the given number of bytes, sets the reference count to 1 and the user count to 0. |
+| __quantum__rt__tuple_add_user   | `void(%Tuple*)` | Increases the current user count by one. |
+| __quantum__rt__tuple_remove_user | `void(%Tuple*)` | Decreases the current user count by one, fails if the user count becomes negative. |
+| __quantum__rt__tuple_copy      | `%Tuple*(%Tuple*, i1)`  | Creates a shallow copy of the tuple if the user count is larger than 0 or the second argument is `true`. |
```

Comment:
Perhaps, could explicitly say here that the calling code is responsible for proper reference counts updates on the contents of the tuple, because the runtime cannot do that?

Reply:
I added a sentence that the reference count of tuple items remains unchanged, and also added a similar sentence to the other copy functions.